src/main/java/com/google/api/codegen/config/ResourceNameMessageConfigs.java

```diff
@@ -134,36 +140,40 @@ private static void loadFieldEntityPairFromResourceReferenceAnnotation(
           "Only one of child_type and type should be set: %s",
           field);
 
-      if (!childType.isEmpty()) {
-        List<ResourceDescriptorConfig> parents =
-            childParentResourceMap.getOrDefault(childType, Collections.emptyList());
-        for (ResourceDescriptorConfig parentResourceDescriptor : parents) {
-          String derivedEntityName = parentResourceDescriptor.getDerivedEntityName();
-          ResourceNameConfig parentResource = resourceNameConfigs.get(derivedEntityName);
-          Preconditions.checkArgument(
-              parentResource != null, "Referencing non-existing parent resource: %s", childType);
-          fieldEntityMap.put(field.getSimpleName(), parentResource.getEntityId());
-        }
+      if (type.equals("*") || childType.equals("*")) {
+        fieldEntityMap.put(field.getSimpleName(), "*");
         continue;
       }
 
-      if (type.equals("*")) {
-        fieldEntityMap.put(field.getSimpleName(), "*");
-        continue;
+      List<ResourceDescriptorConfig> referencedResources;
+      if (!childType.isEmpty()) {
+        referencedResources = childParentResourceMap.get(childType);
+      } else {
+        referencedResources = Collections.singletonList(descriptorConfigMap.get(type));
       }
 
-      String unqualifiedResourceType = ResourceDescriptorConfig.getUnqualifiedTypeName(type);
-      ResourceNameConfig resourceNameConfig =
-          resourceNameConfigs.get(unqualifiedResourceType + "Oneof");
-      if (resourceNameConfig != null
-          && resourceNameConfig.getResourceNameType() == ResourceNameType.ONEOF) {
-        fieldEntityMap.put(field.getSimpleName(), unqualifiedResourceType + "Oneof");
-        continue;
+      for (ResourceDescriptorConfig descriptor : referencedResources) {
+        String unqualifiedName = descriptor.getDerivedEntityName();
+        ResourceNameConfig resource = resourceNameConfigs.get(unqualifiedName);
+
+        switch (resource.getResourceNameType()) {
+          case ANY:
+            fieldEntityMap.put(field.getSimpleName(), "*");
+            break;
+          case SINGLE:
+            fieldEntityMap.put(field.getSimpleName(), unqualifiedName);
+            break;
+          case ONEOF:
+            fieldEntityMap.put(field.getSimpleName(), unqualifiedName);
+            if (((ResourceNameOneofConfig) resource).hasAnyResourceNamePattern()) {
+              fieldEntityMap.put(field.getSimpleName(), "*");
```

Comment:
I think it might be clearer to move the earlier `put` call to an `else` block - so that every branch of this switch statement calls `fieldEntityMap.put` exactly once. (You could even extract this to a method which just returns "the value to put in the map" so you only have a single `fieldEntityMap.put` call.)

Reply:
I might be missing your point Jon, but the former `put` needs to be executed regardless of whether the resource has a `*` pattern, so I can't move it to the else block.

Reply:
You've currently got (pseudo-code)

```text
put (key, value1)
if (condition)
  put (key, value2)
```

I'm suggesting that's better as:

```text
if (condition)
  put (key, value2)
else
  put (key, value1)
```

So it's easy to see that in every path, the key is put exactly once. What's the benefit of putting `unqualifiedName` as the value and *then* putting `"*"` as the value? That's just overwriting it, right?

Reply:
Ahh I see the confusion here. `fieldEntityMap` is a `Multimap`- it supports duplicate keys.

https://github.com/googleapis/gapic-generator/pull/3163/files/74feb00d6b94ff04fed06c784bfcd826529246b3#diff-a0f27b8e4c41d2ebf887482b72314a58R82

Reply:
Ah, that would definitely explain it. Thanks for clarifying.